### PR TITLE
Activity page implementation.

### PR DESCRIPTION
**Pull request by lyw0149**
1. Activity feed
2. Activity request
3. Activity response

@nlpenguin @ethkim @djflexible @ifndefdeadmau5 @SarahhhC @KaylaKwon

### CODE REVIEW

File: src/server/controllers/match.controller.js [hunk 1]
```diff
@@ -9,76 +11,196 @@ import mongoose from 'mongoose';
...
 }
 
-// The mentor accepted the mentoring request
-export function acceptRequest(req, res, next) {
-  // Todo:
-  // update mentor and mentee's request db
+export function getMyActivity(req, res, next) {
+  if (req.session._id) {
+    let activityData = {};
+    //status of match {pending:2, accepted:1, rejected:0}
```
Comment (ifndefdeadmau5) [inline src/server/controllers/match.controller.js]: It would be better use constant rather than 'magic number'.

e.g) const pending = 2

File: src/server/controllers/match.controller.js [hunk 2]
```diff
@@ -9,76 +11,196 @@ import mongoose from 'mongoose';
...
+        res.json(matchCallback.fail);
+      }
+      else {
+        if (mentorDoc.length === 0) {  //if mentor_id is not existing
+          matchCallback.fail.errPoint = 'RequestMentoring - Cannot found mentor.';
+          matchCallback.fail.err = err;
+          res.json(matchCallback.fail);
+        }
+        else {
+          Match.find({mentor_id: matchData.mentor_id, mentee_id: matchData.mentee_id}, (err, matchDoc) => { //find match is existing
```
Comment (ifndefdeadmau5) [inline src/server/controllers/match.controller.js]: Just my 2 cents.
This callback function has almost 50 lines.
To make a code shorter isn't always the best way, but I think this case is either long and hard to read.

My suggestion:

1. Couldn't we move the error handling code to the bottom? 
Because It's better to read if positive case are always comes first than exception handling logic.

2. Isn't there a better way rather than use nested callback? 
It will help your code has better readability.

File: src/server/controllers/users.controller.js [hunk 3]
```diff
@@ -77,7 +79,7 @@ function registerUser(req, res) {
       let cb = authCallback.successRegister;
```
Comment (ifndefdeadmau5) [inline src/server/controllers/users.controller.js]: Not sure what's does 'cb' mean.

Comment (lyw0149) [inline src/server/controllers/users.controller.js]: It means `callback` .

File: src/server/controllers/match.controller.js (hunk 1 above)
Comment (lyw0149) [inline src/server/controllers/match.controller.js]: Nice!! Thank you

File: src/server/controllers/match.controller.js (hunk 2 above)
Comment (lyw0149) [inline src/server/controllers/match.controller.js]: Promise Pattern will solve `callback hell` issue. I'll try to Adjust Promise Pattern.

File: src/server/config/json/auth.callback.js [hunk 4]
```diff
@@ -29,4 +29,8 @@ export default {
       msg: 'Authentication failed. Please sign in first.',
     },
   },
+  fail: {
+    successCode: 0,
+    result: {},
```
Comment (ethkim) [inline src/server/config/json/auth.callback.js]: ah! now i understand why kyungah was putting the eval results inside result{}.

File: src/server/config/json/auth.callback.js [hunk 5]
```diff
@@ -29,4 +29,8 @@ export default {
       msg: 'Authentication failed. Please sign in first.',
     },
   },
+  fail: {
```
Comment (ethkim) [inline src/server/config/json/auth.callback.js]: How does this fail differ from all the other fails?? 
At least provide a comment

File: src/server/controllers/match.controller.js [hunk 6]
```diff
@@ -9,76 +11,196 @@ import mongoose from 'mongoose';
...
-let matchData = new matchSchema();
-let sampleFailResult = {
-  successCode: 0,
-  errPoint: null,
-  err: null,
-};
-
 // Send mentoring request pushing Email to mentor(receiver)
-function sendRequestEmail(res, mentor, mentee, content) {
+function sendRequestEmail(res, mentor, mentee, content, callback) {
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: What does this callback do? Change the name to something more meaningful.

File: src/server/controllers/match.controller.js [hunk 7]
```diff
@@ -9,76 +11,196 @@ import mongoose from 'mongoose';
...
   };
-
   transport.sendMail(mailOptions, function (err, response) {
     if (err) {
-      sampleFailResult.errPoint = 'RequestMentoring - transport.sendMail';
-      sampleFailResult.err = err;
-      res.json(sampleFailResult);
-    } else {
-      res.json('Success');
+      if (typeof callback === 'function')callback(false);
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: move callback(false) to the next line, and surround it by { }

File: src/server/controllers/match.controller.js [hunk 8]
```diff
@@ -9,76 +11,196 @@ import mongoose from 'mongoose';
...
+    }
+    else {
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: move one line up so it reads: "} else {"

File: src/server/controllers/match.controller.js [hunk 9]
```diff
@@ -9,76 +11,196 @@ import mongoose from 'mongoose';
...
+      if (typeof callback === 'function')callback(true);
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: break line before callback() and surround by {}

File: src/server/controllers/match.controller.js (hunk 2 above)
Comment (ethkim) [inline src/server/controllers/match.controller.js]: break line after 100 chars

File: src/server/controllers/match.controller.js [hunk 10]
```diff
@@ -9,76 +11,196 @@ import mongoose from 'mongoose';
...
+    let matchData = req.body;
+    matchData.mentee_id = req.session._id;
+    let match = new Match(matchData);
+    console.log(matchData);
+    User.find({_id: matchData.mentor_id}, (err, mentorDoc) => {  //find mentor_id is existing.
+      if (err) {
+        matchCallback.fail.errPoint = 'RequestMentoring - Error occurred when finding montor_id';
+        matchCallback.fail.err = err;
+        res.json(matchCallback.fail);
+      }
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: move up 1 line

File: src/server/controllers/match.controller.js [hunk 11]
```diff
@@ -9,76 +11,196 @@ import mongoose from 'mongoose';
...
+      else {
+        if (mentorDoc.length === 0) {  //if mentor_id is not existing
+          matchCallback.fail.errPoint = 'RequestMentoring - Cannot found mentor.';
+          matchCallback.fail.err = err;
+          res.json(matchCallback.fail);
+        }
+        else {
```
Comment (ethkim) [inline src/server/controllers/match.controller.js]: move up 1 line

File: src/server/controllers/match.controller.js (hunk 2 above)
Comment (ethkim) [inline src/server/controllers/match.controller.js]: Re: Seongmin's comment:

You don't have to make this nested if-else blocks. For each error handling case, you can just set the required res.json(mathCallback.fail) and return away. That way your code isn't nested (but with a bunch of error handling blocks up top). E.g.

if (error) {
   //do your thing
   return;
}
if (matchDoc.length !== 0) {
    //do your thing
    return;
}

Also, feel free to refactor out callback methods to outside of these methods, if that makes the code less crazy. 

See: http://thecodebarbarian.com/2015/03/20/callback-hell-is-a-myth

File: src/server/routes/match.route.js [hunk 12]
```diff
@@ -4,8 +4,8 @@ import * as match from '../controllers/match.controller';
 const router = express.Router();
 
 router.post('/request', match.requestMentoring);
+router.post('/response', match.responseMentoring);
 
-router.get('/accept', match.acceptRequest);
-router.get('/reject', match.rejectRequest);
+router.get('/activity', match.getMyActivity);
```
Comment (ethkim) [inline src/server/routes/match.route.js]: Leave a comment at the top of each block (line 5 and line 8) that these are the POST/GET methods to make it obvious why we have 2 diff blocks.

File: src/server/models/users.model.js [hunk 13]
```diff
@@ -24,9 +24,4 @@ let userSchema = new Schema({
   },
 });
 
-userSchema.set('toJSON', {
```
Comment (ethkim) [inline src/server/models/users.model.js]: Why was this removed but not for match model?

File: src/server/models/match.model.js [hunk 14]
```diff
@@ -2,22 +2,24 @@ import mongoose from 'mongoose';
...
-  isCancelled: {
-    type: Boolean,
-    default: false,
+  //status of match {pending:0, accepted:1, rejected:2}
+  status: {
+    type: Number,
+    default: 2,
   },
-  cancelDate: Date,
+  response_date: Date,
```
Comment (ethkim) [inline src/server/models/match.model.js]: responseDate? 
I don't know what this date means. Is this the date on which the mentee sent a request? Or was this the date that the mentor accepted/rejected the request? The var name should be obvious which it is, and please provide a comment next to this line. It's not clear why we're storing only one of the 2 dates.

**Comment by ethkim** [comment]
Thanks Yangwoo! Took a first pass. Few comments above.

File: src/server/models/match.model.js (hunk 14 above)
Comment (lyw0149) [inline src/server/models/match.model.js]: time of mentee's request -> request_date,
time of mentor's response -> response_date,

**Comment by ifndefdeadmau5** [comment]
LGTM
